heuristic/delta_calculations.py

```diff
@@ -417,4 +417,127 @@ def regret_objective_function(state, employee, off_shifts, saturdays, days, L_C_
+            3. Partial weekend vil ha noe å si hvis shiftet ble satt på en lørdag eller søndag
+            4. isolated working day er vanskelig å enkelt kalkulere uten å se på dagen før og dagen etter
+            5. Samme som over. Må sjekke dagen før og dagen etter.
+            6. Consecutive days må sjekke strengt tatt L_C_D - 1 dager før og L_C_D -1 dager etter. 
+            7. Negative deviation from contracted hours. Dette går fort å regne ut. Bare ta nåværende - varighet på skift. Alt som blir minus vil da også bryte hard restriksjon.
+            12. daily_rest bør sjekkes. Bare sjekke den dagen. 
+"""
+
+def employee_shift_value(state, e, shift, saturdays, sundays, invalid_shifts, shift_combinations_violating_daily_rest, shift_sequences_violating_daily_rest, shifts_at_week, weeks, shifts_at_day, week, competency_score):
+    day = int(shift[0]/24)
+    daily_rest_error = regret_daily_rest_error(state, day, e, shift, invalid_shifts, shift_combinations_violating_daily_rest, shift_sequences_violating_daily_rest)
+    weekly_rest_error = regret_weekly_rest(state, shifts_at_week, e, week, shift)
+    partial_weekend_error = regret_partial_weekend(state, e, shifts_at_day, saturdays, sundays, day)
+    isolated_days_error = regret_isolated_days(state, e, shifts_at_day, day, weeks)
+    deviation_contracted_hours = regret_deviation_contracted_hours(state, e, shift, week, weeks)
+
+    return (
+            daily_rest_error 
+            + weekly_rest_error 
+            + partial_weekend_error 
+            + isolated_days_error 
+            + deviation_contracted_hours 
+            - 100 * competency_score
             )
+
+def regret_weekly_rest(state, shifts_at_week, e, week, shift):
+    actual_shifts = [(t, v) for t,v in shifts_at_week[week] if state.x[e,t,v] == 1 or (t,v) == (shift[0],shift[1])]
+    off_shift_periods = []
+    important = [7 * 24 * week, 7 * 24 * (week + 1) ]
+  
+    if len(actual_shifts) == 0:
+        off_shift_periods.append(float(important[1] - important[0]))
+        return min(72, off_shift_periods[0])
+
+    else:
+        if(actual_shifts[0][0] - important[0] >= 36):
+            off_shift_periods.append(actual_shifts[0][0] - important[0])
+            
+
+        if(important[1] - (actual_shifts[-1][0] + actual_shifts[-1][1]) >= 36):
+            off_shift_periods.append(important[1] - (actual_shifts[-1][0] + actual_shifts[-1][1]))
+
+        for i in range(len(actual_shifts) - 1):
+            if(actual_shifts[i+1][0] - (actual_shifts[i][0] + actual_shifts[i][1]) >= 36):
+                off_shift_periods.append(actual_shifts[i+1][0] - (actual_shifts[i][0] + actual_shifts[i][1]))
+
+    if len(off_shift_periods) != 0:
+        return min(72, max(off_shift_periods))
+    
+    return -200
+
+
+def regret_partial_weekend(state, e, shifts_at_day, saturdays, sundays, day):
+    if day in saturdays:
+        for t,v in shifts_at_day[day + 1]:
+            if(state.x.get((e, t, v))):
```

Comment:
```python
if (e, t, v) in state.x
```

should be faster

Reply:
This loops seems a litt weird. Since it returns a value regardsless of the `x`-value only the first `(t, v)` is reached. If intentional it would be better to use 
```python
t, v = shifts_at_day[day+1]
```

Reply:
Same applies to sunday-check

Reply:
I don't follow here either. I thought .get((e, t, v)) would be faster than to check if (e, t, v) is a key in x. I don't understand the next comment either. How could i set a list of tuple (t,v) to be t,v? The loop returns a value regardless because I don't check if the next day after a sunday is actually a day in the planning period. If this is not true it would return None, and I would get a valueError later

Reply:
See this answer: https://stackoverflow.com/questions/39582115/python-dictionary-lookup-performance-get-vs-in

The performance difference is small though, so it is probably not necessary. We can look at these things later.

Reply:
For the second comment: 
It should be 
```python
t, v = shifts_at_day[day+1][0]
```

This is equivalent to your current code, but I suspect it might be a bug. 

For each iteration of `(t, v)` in `shifts_at_day[day+1]` you: 
- return -2 if `(e, t, v)` exists in `state.x`
- return 0 if `(e, t, v)` does not exists in `state.x`

Since you return a value regardless of the existence of `(e, t, v)` in `state.x` your loop will always terminate on the first run - and it is, therefore, no need to use a loop.

Reply:
This seems like a good catch. Will check it out. I believe it is a bug then.

Reply:
It is as you said. There is a bug in the code. Instead of checking if (t,v) is present in x I should check if x[t,v] == 1.